### ISSUE TITLE
Typescript allows passing any string as t() argument

### ISSUE THREAD

**Issue by andresespinosapc** [intlify/vue-i18n-next#1116]
### Clear and concise description of the problem

I followed all the steps from the [Typescript support](https://vue-i18n.intlify.dev/guide/advanced/typescript.html#typescript-support) page but it does not give an error when I use the `t` function with a missing key.

It does show autocomplete suggestions in the template though.

### Suggested solution

Typescript should give an error when trying to use a key that is not in the schema

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/intlify/vue-i18n-next/blob/master/.github/CONTRIBUTING.md)
- [X] Read the [Documentation](https://vue-i18n.intlify.dev/)
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

**Comment by OfekA-IAI** [comment]
Any chance this is planned? Has anyone attempted adding this to their project somehow?

**Comment by larrifax** [comment]
Think I got this working by creating my own wrapper composable returning a `t()` function with custom types:
```
type Translations = Messages["en"];

type PathToLeaves<T, Cache extends string = ""> = T extends PropertyKey
  ? Cache
  : {
      [P in keyof T]: P extends string
        ? Cache extends ""
          ? PathToLeaves<T[P], `${P}`>
          : PathToLeaves<T[P], `${Cache}.${P}`>
        : never;
    }[keyof T];

type TranslationPaths = PathToLeaves<Translations>;

export const useTranslation = () => {
  const { t } = useI18n<{ messages: Messages }>();

  return { t: (path: TranslationPaths, named?: NamedValue) => (named ? t(path, named) : t(path)) } as const;
};
```
It's using some dark-magic Typescript tricks that I got from https://stackoverflow.com/a/69881190/852765 (adjusted it a little bit to fit my use-case)

**Comment by Sweet-KK** [comment]
> 认为我通过创建自己的可组合包装器来实现此功能，该包装器可返回`t()`具有自定义类型的函数：
> 
> ```
> type Translations = Messages["en"];
> 
> type PathToLeaves<T, Cache extends string = ""> = T extends PropertyKey
>   ? Cache
>   : {
>       [P in keyof T]: P extends string
>         ? Cache extends ""
>           ? PathToLeaves<T[P], `${P}`>
>           : PathToLeaves<T[P], `${Cache}.${P}`>
>         : never;
>     }[keyof T];
> 
> type TranslationPaths = PathToLeaves<Translations>;
> 
> export const useTranslation = () => {
>   const { t } = useI18n<{ messages: Messages }>();
> 
>   return { t: (path: TranslationPaths, named?: NamedValue) => (named ? t(path, named) : t(path)) } as const;
> };
> ```
> 
> 它使用了我从https://stackoverflow.com/a/69881190/852765获得的一些黑暗魔法 Typescript 技巧（稍微调整一下以适合我的用例）

How can I do to show the key corresponding to the value then typing?
<img width="702" alt="image" src="https://user-images.githubusercontent.com/29600932/198457306-7b1f422f-1771-4dda-bbec-9a4e5bd0e368.png">

**Comment by Vissie2** [comment]
Even though the workaround from @larrifax is somewhat helpful. I'm having a difficult time taking into account all overloads of the `t()` to make usage possible for all parameters. Especially because we can't just make a type of the function overloads due to a design limitation of TypeScript (https://github.com/microsoft/TypeScript/issues/32164).

This also a problem for other translation functions like `tm()` and `rt()`.

**Comment by OfekA-IAI** [comment]
> Even though the workaround from @larrifax is somewhat helpful. I'm having a difficult time taking into account all overloads of `t()` to make usage possible for all parameters. Especially because we can't just make a type of the function overloads due to a design limitation of TypeScript ([microsoft/TypeScript#32164](https://github.com/microsoft/TypeScript/issues/32164)).
> 
> This is also a problem for other translation functions like `tm()` and `rt()`.

Agree, it's working but a pain to do and configure. This should be built in 🙏🙂

**Comment by isbosykh** [comment]
Any updates?

**Comment by danwithabox** [comment]
I too would like this feature!
I made my own `useI18nStrict()` composable that returns a stricter `t()` not unlike @larrifax's, but then I realized `tm()` would be a lot more hassle because of the return type.
A fully signature-compatible `useI18nStrict()` as a drop-in replacement is about the best we could do besides official support, but mimicking the entire surface area of `useI18n()` seems painful 😅

**Comment by loic-brtd** [comment]
I had the same need of type safety for i18n paths.

As a workaround, I copied theses types from [vue-i18n/packages/core-base/src/types/utils.ts](https://github.com/intlify/vue-i18n/blob/v9.9.1/packages/core-base/src/types/utils.ts) in my codebase, and I replaced  the types of some methods (here `t` and `tm`) in the `export` clause :

```ts
import { createI18n, type NamedValue } from "vue-i18n";
import { messages } from "./fr";

// ---- Taken from https://github.com/intlify/vue-i18n/blob/v9.9.1/packages/core-base/src/types/utils.ts
type __ResourcePath<T, Key extends keyof T> = Key extends string
  ? T[Key] extends Record<string, any>
    ?
        | `${Key}.${__ResourcePath<T[Key], Exclude<keyof T[Key], keyof any[]>> &
            string}`
        | `${Key}.${Exclude<keyof T[Key], keyof any[]> & string}`
    : never
  : never;
type _ResourcePath<T> = __ResourcePath<T, keyof T> | keyof T;
type ResourcePath<T> = _ResourcePath<T> extends string | keyof T
  ? _ResourcePath<T>
  : keyof T;

type ResourceValue<
  T,
  P extends ResourcePath<T>
> = P extends `${infer Key}.${infer Rest}`
  ? Key extends keyof T
    ? Rest extends ResourcePath<T[Key]>
      ? ResourceValue<T[Key], Rest>
      : never
    : never
  : P extends keyof T
  ? T[P]
  : never;
// ----

type MessageSchema = typeof messages.fr;

// Currying to simplify the types
export type MessagePath = ResourcePath<MessageSchema>;
export type MessageValue<Path extends ResourcePath<MessageSchema>> =
  ResourceValue<MessageSchema, Path>;

const i18n = createI18n<[MessageSchema], "fr">({
  legacy: false,
  warnHtmlMessage: false,
  locale: "fr",
  fallbackLocale: "fr",
  formatFallbackMessages: true,
  messages,
});

// Don't forget to `Omit` the properties to replace them completely (no overloading)
type I18nType = typeof i18n;
type PatchedI18nType = Omit<I18nType, "global"> & {
  global: Omit<I18nType["global"], "t" | "tm"> & {
    t<Path extends MessagePath>(
      path: Path,
      named?: NamedValue | string[]
    ): string;

    tm<Path extends MessagePath>(
      path: Path,
      named?: NamedValue | string[]
    ): MessageValue<Path>;
  };
};

export default i18n as PatchedI18nType;
```

This works for all usages of `i18n.global.t(...)` and `i18n.global.tm(...)` but it doesn't work when using `$t(...)` in templates for example.

**Comment by kazupon** [comment]
Thank you for the report and the great discussion!

This is a duplicate of #1195 and #2307.

In v12 (still alpha), we've improved the type signatures by replacing `Key extends string` with `ResourceKeys | (string & {})` in all `$t` and `t()` overloads. This change makes IDE autocompletion work much better — resource keys defined via `DefineLocaleMessage` are now shown as prioritized suggestions.

However, **strict type errors for invalid keys** are not yet implemented. The `string & {}` technique improves completion but still accepts any string, because `t(key, defaultMsg)` overloads require accepting arbitrary strings as keys.

Full type safety (errors on invalid keys) is tracked in #2307 and would require an opt-in strict mode to avoid breaking the `defaultMsg` overloads.

Closing this as a duplicate of #2307.
Thanks!
